javaslang/src/main/java/javaslang/collection/JavaConverters.java

```diff
@@ -0,0 +1,437 @@
+ * </ul>
+ *
+ * Subtypes of the Javaslang types mentioned above can have special views that make use of optimized implementations.
+ *
+ * @author Daniel Dietrich, Pap Lőrinc, Sean Flanigan
+ * @since 2.1.0
+ */
+class JavaConverters {
+
+    private JavaConverters() {
+    }
+
+    static <T> java.util.List<T> asJava(IndexedSeq<T> seq, ChangePolicy changePolicy) {
+        return new ListView<>(seq, changePolicy.isMutable());
+    }
+
+    static <T> java.util.List<T> asJava(LinearSeq<T> seq, ChangePolicy changePolicy) {
+        return new ListView<>(seq, changePolicy.isMutable());
+    }
+
+    static <T> java.util.Set<T> asJava(Set<T> set, ChangePolicy changePolicy) {
+        return TODO("new SetView<>(set, changePolicy.isMutable());");
+    }
+
+    static <T> java.util.NavigableSet<T> asJava(SortedSet<T> set, ChangePolicy changePolicy) {
+        return TODO("new NavigableSetView<>(set, changePolicy.isMutable());");
+    }
+
+    static <K, V> java.util.Map<K, V> asJava(Map<K, V> map, ChangePolicy changePolicy) {
+        return TODO("new MapView<>(map, changePolicy.isMutable());");
+    }
+
+    static <K, V> java.util.NavigableMap<K, V> asJava(SortedMap<K, V> map, ChangePolicy changePolicy) {
+        return TODO("new NavigableMapView<>(map, changePolicy.isMutable());");
+    }
+
+    /*
+    static <K, V> java.util.Map<K, java.util.Collection<V>> asJava(Multimap<K, V> map, ChangePolicy changePolicy) {
+        return TODO("new MapView<>(map, changePolicy.isMutable());");
+    }
+
+    TODO: create interface javaslang.collection.SortedMultimap
+    static <K, V> java.util.NavigableMap<K, java.util.Collection<V>> asJava(SortedMultimap<K, V> map, ChangePolicy changePolicy) {
+        return TODO("new NavigableMapView<>(map, changePolicy.isMutable());");
+    }
+    */
+
+    enum ChangePolicy {
+
+        IMMUTABLE, MUTABLE;
+
+        boolean isMutable() {
+            return this == MUTABLE;
+        }
+    }
+
+    // -- private view implementations
+
+    /**
+     * Encapsulates the access to delegate and performs mutability checks.
+     *
+     * @param <C> The Javaslang collection type
+     */
+    private static abstract class HasDelegate<C extends Traversable<?>> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        private C delegate;
+        private final boolean mutable;
+
+        HasDelegate(C delegate, boolean mutable) {
+            this.delegate = delegate;
+            this.mutable = mutable;
+        }
+
+        boolean isMutable() {
+            return mutable;
+        }
+
+        C getDelegate() {
+            return delegate;
+        }
+
+        C setDelegate(C newDelegate) {
+            if (mutable) {
+                final C previousDelegate = delegate;
+                delegate = newDelegate;
+                return previousDelegate;
+            } else {
+                throw new UnsupportedOperationException();
```

Comment:
This incurs a fair bit of unnecessary work building the newDelegate collection in the immutable case.

What about another method `getDelegateForMutation()` which can throw the exception earlier in the immutable case, which will be called by the add/remove/mutator methods instead of `getDelegate()`?